src/testSession.ts

```diff
@@ -0,0 +1,234 @@
+        }
+      }
+    }
+  }
+};
+
+/**
+ * Represents a test session, which is a unique location for non-unit test (nut)
+ * artifacts such as a project and a mocked home dir.  It also provides easy
+ * access to org usernames created by setup commands, cwd stubbing, and a way to
+ * zip up the test session.
+ *
+ * Create a TestSession instance with: `const testSession = TestSession.create(options)`
+ *
+ * Fine-grained control over certain test session details are provided by these
+ * environment variables:
+ *   TESTKIT_SESSION_DIR = Overrides the default directory for the test session
+ *   TESTKIT_HOMEDIR = path to a home directory that the tests will use as a stub of os.homedir
+ *   TESTKIT_ORG_USERNAME = an org username to use for test commands. tests will use this org rather than creating new orgs.
+ *   TESTKIT_PROJECT_DIR = a SFDX project to use for testing. the tests will use this project directly.
+ *   TESTKIT_SAVE_ARTIFACTS = prevents a test session from deleting orgs, projects, and test sessions.
+ *   TESTKIT_ZIP_SESSION = zips the session dir when this is true
+ */
+export class TestSession {
+  public id: string;
+  public createdDate: Date;
+  public dir: string;
+  public homeDir: string;
+  public project?: TestProject;
+  public orgUsername = env.getString('TESTKIT_ORG_USERNAME');
+  public sandbox = createSandbox();
+
+  private debug: Debugger;
+  private cwdStub?: SinonStub;
+  private overridenDir?: string;
+
+  private constructor(options: TestSessionOptions = {}) {
+    this.debug = debug('testkit:session');
+    this.createdDate = new Date();
+    this.id = genUniqueString(`${this.createdDate.valueOf()}%s`);
+
+    // Create the test session directory
+    this.overridenDir = env.getString('TESTKIT_SESSION_DIR') || options.sessionDir;
+    this.dir = this.overridenDir || path.join(process.cwd(), `test_session_${this.id}`);
+    fsCore.mkdirpSync(this.dir);
+
+    // Write the test session options
+    fsCore.writeJsonSync(path.join(this.dir, 'testSessionOptions.json'), JSON.parse(JSON.stringify(options)));
+
+    // Setup a test project and stub process.cwd to be the project dir
+    if (options.project) {
+      let projectDir = env.getString('TESTKIT_PROJECT_DIR');
+      if (!projectDir) {
+        this.project = new TestProject({ destinationDir: this.dir, ...options.project });
+        projectDir = this.project.dir;
+      }
+
+      // the default bin/run in execCmd will no longer resolve properly when
+      // a test project is used since process.cwd is changed.  If the
+      // TESTKIT_EXECUTABLE_PATH env var is not being used, then set it
+      // to use the bin/run from the cwd now.
+      env.setString('TESTKIT_EXECUTABLE_PATH', path.join(process.cwd(), 'bin', 'run'));
+
+      this.stubCwd(projectDir);
+    }
+
+    // Create the .sfdx directory used by this test
+    this.homeDir = env.getString('TESTKIT_HOMEDIR', this.dir);
```

Comment:
@shetzel test session is always moving home dir to a temporary location. This is a good thing.
 
But I don't see it being applied to any of the exec commands in `setupCommands`.

There are a number of other env vars that are referenced when trying to resolve where data related to oclif/dx that may break things when home is moved.

These are the env vars that are applied in Jenkins for toolbelt.
```
            "HOME=${home}",
            "APPDATA=${home}",
            "USERPROFILE=${home}",
```

Reply:
In my most recent changes that I haven't yet pushed the `HOME` env var is set as you have above.  I don't have `APPDATA` or `USERPROFILE` set as env vars.  Do you think they should be for testkit?  I don't think we necessarily have to match Jenkins env unless it prevents testkit use within a Jenkins env.

Reply:
Running tests locally can depend on existing installation of sfdx. Moving home during a test can disrupt finding where things are installed.

These env vars will probably need to be set to original HOME before it is changed in the test.